tools/chain-util/pkg/export/export.go

```diff
@@ -0,0 +1,173 @@
+package export
+
+import (
+	"context"
+	"io"
+	"path/filepath"
+
+	bserv "github.com/ipfs/go-blockservice"
+	cid "github.com/ipfs/go-cid"
+	badgerds "github.com/ipfs/go-ds-badger"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	offline "github.com/ipfs/go-ipfs-exchange-offline"
+	format "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	dag "github.com/ipfs/go-merkledag"
+	errors "github.com/pkg/errors"
+
+	block "github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	chain "github.com/filecoin-project/go-filecoin/internal/pkg/chain"
+	encoding "github.com/filecoin-project/go-filecoin/internal/pkg/encoding"
+)
+
+var log = logging.Logger("chain-util/export")
+
+// NewChainExporter returns a ChainExporter.
+func NewChainExporter(repoPath string, out io.Writer) (*ChainExporter, error) {
+	// open the datastore in readonly mode
+	badgerOpt := &badgerds.DefaultOptions
+	badgerOpt.ReadOnly = true
+
+	log.Infof("opening filecoin datastore: %s", repoPath)
+	badgerPath := filepath.Join(repoPath, "badger/")
+	log.Infof("opening badger datastore: %s", badgerPath)
+	badgerDS, err := badgerds.NewDatastore(badgerPath, badgerOpt)
```

Comment:
Since we are not using the go-filecoin/repo abstraction there doesn't seem to be any protection against multiple processes obtaining a handle to a datastore wrapping the same path.

I'm a bit worried that this could lead to accidental corruption if someone runs this tool while a daemon is running.

Lowest effort solution is to include an all CAPs user facing warning somewhere.

I think "the right thing" here is to open up a go-filecoin repo abstraction and then grab the datastores with `Datastore` and `ChainDatastore`.  Even if we're lucky and concurrent read-read and read-write works for badger right now it adds a hidden requirement to our datastore choice that could erode with time.

Reply:
> Since we are not using the go-filecoin/repo abstraction there doesn't seem to be any protection against multiple processes obtaining a handle to a datastore wrapping the same path.

Since we are using badger to open the repo it will fail to open if a [repo lock exists](https://github.com/dgraph-io/badger#opening-a-database), from the linked README.md:
> Please note that Badger obtains a lock on the directories so multiple processes cannot open the same database at the same time.

Here is what a user will see if they attempt to run this binary on a repo that currently has a lock held by another (go-filecoin) process.
```
./chain-util export --repo=/home/frrist/.filecoin/ --out=/tmp/test.car
2019-11-11T11:25:10.586-0800    INFO    chain-util/export       export/export.go:44     opening repo: /home/frrist/.filecoin/
2019-11-11T11:25:10.586-0800    WARN    chain-util      chain-util/main.go:75   Cannot acquire directory lock on "/home/frrist/.filecoin/badger".  Another process is using this Badger database.: resource temporarily unavailable
```

Also note that we are opening the repo in read-only mode:
https://github.com/filecoin-project/go-filecoin/pull/3594/files#diff-2bbea98c52d3e94072ef6b7ba1a9250dR27-R29

Reply:
That's good!  It moves this from a correctness to a bitrot thing. It still doesn't address this part of the comment:

>  Even if we're lucky and concurrent read-read and read-write works for badger right now it adds a hidden requirement to our datastore choice that could erode with time.

Can we assume that all ipfs datastore implementations MUST repo lock?  If we can, great.  If not the moment we pivot to another datastore or we open up user configured datastores we/users are going to need to satisfy a hidden requirement for interop with this tool which means there's some odds somebody *won't* satisfy this requirement and they'll be left wondering why the tool is corrupting their datastore.

Reply:
Grabbing the filecoin repo lock enforces safety across all future datastore implementations.

Reply:
I think I might have misunderstood the initial comment, let me try again.

> Since we are not using the go-filecoin/repo abstraction

Meaning since we are not calling the go-filecoin method [OpenFSRepo](https://github.com/filecoin-project/go-filecoin/blob/master/internal/pkg/repo/fsrepo.go#L131), correct?

> doesn't seem to be any protection against multiple processes obtaining a handle to a datastore wrapping the same path

There _is_ protection against multiple processes getting a handle to a _datastore_  (as I have shown this is provided by the badger datastore package), however since go-filecoin [creates is own repo lock](https://github.com/filecoin-project/go-filecoin/blob/master/internal/pkg/repo/fsrepo.go#L164) and since we are not checking for it to exists here, we are left open to possible corruption in the event that we change our datastore/repo implementation? 

Is the ask that a check for filecoin repo lock is added? (as is done [here](https://github.com/filecoin-project/go-filecoin/blob/master/internal/pkg/repo/fsrepo.go#L164))

Reply:
Yeah this is all exactly what I'm thinking.  If we use openfsrepo we get the guarantee of go-filecoin repo locking preventing contention with other software that opens the repo (go-filecoin, the migration tool...) regardless of what datastore we use.

Originally I wasn't sure if badger had its own protection which is why I blocked approval and why I approved when you told me that it does.